benchmarks/locks/clh_mutex.h

```diff
@@ -0,0 +1,169 @@
+ * a register, or at worse, on the current thread's stack.
+ *
+ * This locking mechanism has some similarities with the MCS lock because each
+ * thread is spinning on the "islocked" of the previous node, which means that
+ * each thread is waiting on a different cache-line, thus reducing
+ * synchronization.
+ * http://www.cise.ufl.edu/tr/DOC/REP-1992-71.pdf
+ * The main different from the MCS lock (section 2) is that MCS may have to do
+ * an Compare-And-Swap in addition to the atomic_store() in the unlock function.
+ *
+ * The CLH lock uses the (theoretically) minimum amount of synchronization
+ * operations to perform a mutual exclusion lock, because:
+ * - In the lock() function it has a single synchronized operation
+ *   atomic_exchange() with one acquire and one release barrier.
+ * - In the unlock() function is has a single synchronized operation, the
+ *   atomic_store() with a release barrier.
+ */
+
+#include <stdatomic.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <pthread.h>
+#include <sched.h>
+
+
+typedef struct clh_mutex_node_ clh_mutex_node_t;
+
+struct clh_mutex_node_
+{
+    atomic_int succ_must_wait;
+};
+
+typedef struct
+{
+    clh_mutex_node_t * mynode;
+    int padding[64];  // To avoid false sharing with the tail
+    _Atomic (clh_mutex_node_t *) tail;
+} clh_mutex_t;
+
+static clh_mutex_node_t * clh_mutex_create_node(int islocked)
+{
+    clh_mutex_node_t * new_node = (clh_mutex_node_t *)malloc(sizeof(clh_mutex_node_t));
+    atomic_store_explicit(&new_node->succ_must_wait, islocked, memory_order_relaxed);
+    return new_node;
+}
+
+
+/*
+ * Initializes the mutex, creating a sentinel node.
+ *
+ * Progress Condition: Wait-Free Population Oblivious
+ */
+void clh_mutex_init(clh_mutex_t * self)
+{
+    // We create the first sentinel node unlocked, with islocked=0
+    clh_mutex_node_t * node = clh_mutex_create_node(0);
+    self->mynode = node;
+    atomic_store(&self->tail, node);
+}
+
+
+/*
+ * Destroy the mutex, clearing all memory.
+ * You must be sure that there are no other threads currently holding
+ * the lock or attempting to.
+ *
+ * Progress Condition: Wait-Free Population Oblivious
+ */
+void clh_mutex_destroy(clh_mutex_t * self)
+{
+    free(atomic_load(&self->tail));
+}
+
+
+/*
+ * Locks the mutex for the current thread. Will wait for other threads
+ * that did the atomic_exchange() before this one.
+ *
+ * Progress Condition: Blocking
+ */
+void clh_mutex_lock(clh_mutex_t * self)
+{
+    // Create the new node locked by default, setting islocked=1
+    clh_mutex_node_t *mynode = clh_mutex_create_node(1);
+    clh_mutex_node_t *prev = atomic_exchange(&self->tail, mynode);
+
+    // This thread's node is now in the queue, so wait until it is its turn
+    // The original benchmark had a relaxed load, but this violates mutual exclusion
+    int prev_islocked = atomic_load_explicit(&prev->succ_must_wait, memory_order_relaxed);
```

Comment:
It's still relaxed here...
Why don't you something like `#IFDEF FAIL` to have both versions in code?